### PR TITLE
mgr/cephadm: speed up upgrade when using agent

### PR DESCRIPTION
**Pull request by adk3798**
Returning completely from the upgrade procedure when
any hosts at all don't have their metadata up-to-date
is unnecessarily strict. We can just skip daemons
specifically on hosts where the metadata is not up-to-date
and then just make sure the metadata is up-to-date on all
hosts before calling the upgrade complete

Fixes: https://tracker.ceph.com/issues/53598

Signed-off-by: Adam King <adking@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/upgrade.py [hunk 1]
```diff
@@ -610,6 +603,14 @@ def _do_upgrade(self):
             for d in daemons:
                 if d.daemon_type != daemon_type:
                     continue
+                # if we don't have up-to-date metadata from this daemon's host we don't know
+                # for sure if the currently known image digest for this daemon is correct
+                # so we can't say for sure if it needs upgrade or if it's been upgraded (done)
+                if (
+                    self.mgr.use_agent
+                    and d.hostname in [h for h in self.mgr.cache.get_hosts() if not self.mgr.cache.host_metadata_up_to_date(h)]
+                ):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/upgrade.py:612]:
```suggestion

    def agent_has_up_to_date_info(hostname: str): -> bool
        """
        if we don't have up-to-date metadata from this daemon's host we don't know
        for sure if the currently known image digest for this daemon is correct
        so we can't say for sure if it needs upgrade or if it's been upgraded (done)
        """
        return (
                    self.mgr.use_agent
                    and d.hostname in [h for h in self.mgr.cache.get_hosts() if not self.mgr.cache.host_metadata_up_to_date(h)]
                )
```

imo do_upgrade() is already complicated enough.

File: src/pybind/mgr/cephadm/upgrade.py [hunk 2]
```diff
@@ -694,7 +694,7 @@ def _do_upgrade(self):
                 to_upgrade.append(d_entry)
 
                 # if we don't have a list of others to consider, stop now
-                if not known_ok_to_stop:
+                if not known_ok_to_stop and d.daemon_type in ['osd', 'mds', 'mon']:
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/upgrade.py:697]: can you change the order of the expression?

```suggestion
                if d.daemon_type in ['osd', 'mds', 'mon'] and not known_ok_to_stop:
```

cause checking for `known_ok_to_stop` only makes if we're talking about osds, mdss or mons. And *not* vice versa.

File: src/pybind/mgr/cephadm/agent.py [hunk 3]
```diff
@@ -356,6 +356,12 @@ def _request_agent_acks(self, hosts: Set[str], increment: bool = False) -> None:
                 host, self.mgr.cache.agent_ports[host], {'counter': self.mgr.cache.agent_counter[host]}, self.mgr)
             message_thread.start()
 
+    def _request_ack_all_not_up_to_date(self) -> None:
+        self.mgr.agent_helpers._request_agent_acks(
+            set([h for h in self.mgr.cache.get_hosts() if
```
[the remark this answers is not in the document]
Comment (adk3798) [reply src/pybind/mgr/cephadm/agent.py:361]: This wasn't working for me

`Argument 1 to <set> has incompatible type "List[str]"; expected "str"`

[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/agent.py:361]: Fascinating! `set([1,2,3])` should be equal to `{1,2,3}`

File: src/pybind/mgr/cephadm/upgrade.py (hunk 1 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/upgrade.py]: realized I didn't need all this and we already have a function to check if a host's metadata is up to date in the host cache

File: src/pybind/mgr/cephadm/agent.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply src/pybind/mgr/cephadm/agent.py:361]: Do you need an inner list comprehension? Would directly using a [set comprehension](https://python-reference.readthedocs.io/en/latest/docs/comprehensions/set_comprehension.html) possibly be more succinct and pass the type check?

**Review by sebastian-philipp — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2022-01-03_15:37:23-orch:cephadm-wip-swagner-testing-2022-01-03-1230-distro-default-smithi/

* Failures are due to the environment

**Pull request merged**